src/installer/records.py

```diff
@@ -130,8 +145,7 @@ def from_elements(cls, path, hash_, size):
 
 def parse_record_file(rows):
     # type: (Iterator[str]) -> Iterator[Record]
-    """Parse a RECORD file, provided as an iterator of record lines.
-    """
+    """Parse a RECORD file, provided as an iterator of record lines."""
     reader = csv.reader(rows, delimiter=",", quotechar='"', lineterminator=os.linesep)
```

Comment:
`lineterminator` should not depend on the OS on which the library is used! The wheel library always uses `\n` as the line terminator, regardless of the operating system.

Reply:
This is also a detail that should be added to the wheel spec.

Just to be clear, all I'm saying is that these questions come up whenever anyone works on a wheel implementation, and we should capture them and make sure there's a clear answer recorded somewhere - and the spec is probably the best place.

Reply:
https://www.python.org/dev/peps/pep-0376/#record

PEP 376 says use os.linesep.

Reply:
@pfmoore can/should I move things over to packaging.python.org in the process?

Reply:
PEP 427 (the wheel spec) simply defers to PEP 376 (which defines `.dist-info` directories) on `RECORD`, but that spec can afford to be platform-dependant (IMO it does not need to, but that’s another issue), while the wheel format shouldn’t be. From what I can read from the wheel spec, there are two things in `RECORD` that may need clarification.

The first is `lineterminator`:

> The `RECORD` file is a CSV file, composed of records, one line per installed file. The `csv` module is used to read the file, with these options:
>
> * field delimiter : `,`
> * quoting char : `"`
> * line terminator : `os.linesep` (so `\r\n` or `\n`)

and the other is the third path element variant:

> the file's **path**
> * a `'/'`-separated path, relative to the base location, if the file is under the base location.
> * a `'/'`-separated path, relative to the base location, if the file is under the installation prefix AND if the base location is a subpath of the installation prefix.
> * an absolute path, using the local platform separator.

Both values should be interpreted in a platform-dependent way, which does not work well for wheels.

My proposal is to adding the followings to PEP 427 (item 4 in [this section](https://www.python.org/dev/peps/pep-0427/#the-dist-info-directory)):

> The `RECORD` file […]. Unlike PEP376 […]. The file should always be read using filed delimiter `,`, quoting char `"`, and line terminator `\n`. The **path** component should always use `'/'` as the separator, instead of the local platform path separator.

I can draft a pull request to the PEP repository if this makes sense.

Reply:
I'll change lineterminator to "\n". :)

Reply:
@pfmoore looking at https://discuss.python.org/t/refine-wheel-spec-to-designate-directory-entries/1667 and https://github.com/python/peps/pull/1045, I wonder if we should make these changes in-place to wheel 1.0, or if they should be in a wheel 1.1.

I'm 100% OK either way, with a slight preference toward the former since we're amending a standard based on current "relevant" practice in all the tools we know of.

Reply:
> PEP 427 (the wheel spec) simply defers to PEP 376 (which defines .dist-info directories) on RECORD

Actually, it doesn't. At no point that I could find does it say that the RECORD file in the wheel follows the format of the PEP 376 RECORD file. (It *nearly* says it in a few places, but never actually comes out and does so). And PEP 376 is problematic in other ways - for example, it allows absolute filenames, which make no sense in a wheel. Longer term, I think PEP 427 is badly under-specified and needs a revision to fix the various "open to interpretation" problems it has. That rewrite would need to be a "Wheel 1.1" spec.

However, in the interest of not derailing momentum on *this* project, I think we can take the following view:

* `RECORD` should use `,` as delimiter, `"` as quote, and `\n` as line separator.
* `RECORD` should use `/` for paths, which must all be relative to the wheel root.

I'd be OK with these two changes being published as an update to the Wheel 1.0 spec. But they would need a note covering compatibility. Something like "the spec was never intended to allow platform-dependent line and path separators in the `RECORD` file, but the original wording did not make that clear. As no known tools ever took advantage of the ability to do this, the spec has now been changed to explicitly disallow it".

The change should be posted to the Packaging Discourse before being submitted, just to make sure the packaging community has a chance to comment. I don't expect it to be controversial, but hey, this is packaging 🙄 

I'm trying really hard to avoid needing a Wheel 1.1 here, because if we decide we need Wheel 1.1 to clear up the vagueness in 1.0, we have a hard decision to make about how this project will handle files that declare they follow Wheel 1.0 (i.e., every existing wheel...)

By the way, while we're discussing `RECORD` files, in theory when this project installs a wheel, it should write a `RECORD` file in PEP 376 format, *not* in PEP 427 format. In other words, os.linesep and the weird `/`-if-relative-else-platform-native filenames. I'm betting we won't do that. But if not, we should probably tidy up PEP 376 to match what tools really do, as well.

(Sigh. Remember the good old days when PEPs could get away with being hand-wavey and vague? I think I need to go and give past me a really good talking to...)

Reply:
> @pfmoore can/should I move things over to packaging.python.org in the process?

I think that making packaging.python.org the canonical spec itself needs a PEP - see how the core metadata specs did it.

Reply:
> I'm trying really hard to avoid needing a Wheel 1.1 here,

I agree -- let's just amend wheel 1.0 based on "what actually happens" with the suggested wording for the note.

> I think that making packaging.python.org the canonical spec itself needs a PEP - see how the core metadata specs did it.

I don't think we don't *need* a PEP -- it's just been a convenient occasion to make the switch to packaging.python.org along with a PEP for updates (I remember @ncoghlan has stated so in a bunch of places, but I'll need to go digging to find one of these statements).

FWIW, we're making progress on this project + wheel's API toward better quality implementation / tooling for handling + installing wheels; as well as amending the wheel 1.0 PEP. I think that's as an time as any, to make the switch. :)

Reply:
If we move the spec to packaging.python.org (I'd be interested if you could find what @ncoghlan said - he was on the SC so has a better idea of how far we can mangle the standard PEP processes than I do 😉) then I'd prefer if we tidied it up as part of that exercise, and didn't just copy it over as is. There's a chunk of tidying up needed¹ and the move would be the best time to do that.

If you're thinking of just moving the text across unchanged (or with the sort of minor tweaks we've discussed here) I don't really see the point, TBH. It's not like a PR against the PEP is any harder than a PR against the packaging docs... (most of this discussion is about moving the docs, not about the changes which I already said were OK as a PEP modification)

¹ For example, I just spotted that the only discussion of hashes in `RECORD` is in the "Signed wheels" section, even though signed wheels are hardly ever used!

Reply:
> For example, I just spotted that the only discussion of hashes in RECORD is in the "Signed wheels" section, even though signed wheels are hardly ever used!

Maybe that's because without signing, `RECORD` does not protect against tampering, as anybody who wanted to modify the wheel could simply edit `RECORD` to match.

Reply:
https://github.com/pradyunsg/installer/pull/20 covers this. All the other discussion is preserved, but as @pfmoore noted, isn't 100% on-topic for this project/PR. :)